Feature/#438537 レンタル機能

aizawa-being opened the pull request.

aizawa-being committed:
#438537 レンタル機能

貸出･返却機能の実装。

aizawa-being committed:
#438537 レンタル機能

テストプログラムの作成。ロジックの共通化。

aizawa-being committed:
#438537 レンタル機能

名称変更

aizawa-being committed:
#438537 レンタル機能

コメントのリファクタリング。

aizawa-being committed:
#438537 レンタル機能

不要なusingの削除

k-kobayashi-being commented on Libra/Views/LibraForm.cs:
```diff
@@ -46,7 +46,32 @@ public partial class LibraForm : Form {
         /// <param name="sender"></param>
         /// <param name="e"></param>
         private void Borrow_Click(object sender, EventArgs e) {
+            if (booksDataGridView.SelectedCells.Count == 0) {
+                return;
+            }
+            // 選択されている行を取得
+            DataGridViewCell wSelectedCell = this.booksDataGridView.SelectedCells[0];
+            DataGridViewRow wSelectedRow = wSelectedCell.OwningRow;
+
+            // 貸出する書籍のIDを取得
+            int wBookId = (int)wSelectedRow.Cells["bookIdColumn"].Value;
 
+            // 貸出処理
+            this.FLibraControl.BorrowBook(this.userNameTextBox.Text, wBookId);
+
```
BorrowBook()にthis.userNameTextBox.Textを渡してそっちのメソッドで利用者名の有無を判定するのは回りくどい気がしました。
ここでは2つ指摘があります。
①貸出中エラー＞利用者名未入力エラー
→個人的な感覚ですが、エラーの表示として優先度はこの方が親切かなって思いました。
　なぜなら同じ行の選択で2種類のエラーが出ないからです。

②BorrowBookの処理内容
→BorrowBookは**貸出処理**なので利用者名の判定が中にあるのは違和感があります。
```
　if (string.IsNullOrWhiteSpace(vUserName)) {
                this.FMessageBoxService.Show(MessageTypeEnum.UserNameNotInput);
                return;
            }
```
　この判定はここにあってもいいかと思います。

修正するなら
・まず貸出中を判定する。
　57行目までで貸出する書籍の情報は取得するので、このタイミングで貸出中か判定する。
・BorrowBook()を呼び出す前に利用者名が入っているかを判定する。
でいいかと思います。

①は必須ではないですが修正したほうが良いかと。
②は必須の修正対象とします。

aizawa-being replied:
①貸出中エラー＞利用者名未入力エラー
BorrowBookメソッド呼び出し前に、DataGridViewの利用者名にデータがあるか判定する処理を追加しました。
ただし、DB更新時の貸出中判定処理は別途必要であるため、修正していません。
⇒内部処理的には確認している対象が違いますが、外見的には貸出中の判定が2回行われているように見えるようになっています。

②BorrowBookの処理内容
利用者名の入力判定を貸出処理前に判定するよう修正しています。

aizawa-being committed:
Merge branch 'develop' into feature/#438537_レンタル機能

# Conflicts:
#	Libra/Controls/ILibraBookService.cs
#	Libra/Controls/ILibraControl.cs
#	LibraUnitTest/LibraUnitTest.csproj

aizawa-being committed:
Marge内容反映。

aizawa-being committed:
#438537 レンタル機能

①貸出中エラー＞利用者名未入力エラー
DBアクセス前に利用者名の入力があるか判定する処理を追加。
②BorrowBookの処理内容
利用者名の入力判定を貸出処理前に判定するよう修正。

aizawa-being committed:
#438537 レンタル機能

検索機能で行ったリファクタリング内容をレンタル機能の実装にも反映。

k-kobayashi-being commented on Libra/Views/LibraForm.cs:
```diff
@@ -46,7 +46,49 @@ public partial class LibraForm : Form {
+            DataGridViewRow wSelectedRow = wSelectedCell.OwningRow;
+
+
+            // 貸出する書籍の情報を取得
```
無駄な改行が2行続いています。
修正をお願いします。

aizawa-being replied:
余分な空白行を削除いたしました。

k-kobayashi-being commented on Libra/Views/LibraForm.cs:
```diff
@@ -46,7 +46,49 @@ public partial class LibraForm : Form {
+
+
+            // 貸出する書籍の情報を取得
+            int wBookId = (int)wSelectedRow.Cells["bookIdColumn"].Value;
+            var wTitle = (string)wSelectedRow.Cells["titleColumn"].Value;
+            var wUserName = wSelectedRow.Cells["userNameColumn"].Value;
+
+            IMessageBoxUtil wMessageBox = new MessageBoxUtil();
+
+            // 既に貸出中
+            if (wUserName != DBNull.Value) {
+                wMessageBox.Show(MessageTypeEnum.AlreadyBorrowed, wTitle);
+                return;
+            }
+
+            // 利用者名は入力が必須
+            if (string.IsNullOrWhiteSpace(this.userNameTextBox.Text)) {
+                wMessageBox.Show(MessageTypeEnum.UserNameNotInput);
+                return;
+            }
+
+            // 貸出処理
+            this.FLibraControl.BorrowBook(wBookId, this.userNameTextBox.Text);
+
```
これは修正依頼ではなく今後意識して欲しいことです。
59行目のwTitleについて
変数に取ることはOK。
ただ、どう使うから変数に取る。
や、どこで使うために変数に取る。
という感覚を磨ければいいなと思いました。
今回の場合、65行目のif文に入らない場合、wTitleは使わない変数ですよね？
であればwTitleはその中で取得するようにすればいい。
そうなるとメッセージボックスへ引数として渡すためだけにわざわざ変数に取る必要はあるのか？
という思考になっていくと思います。
とっても細かいことだし、可読性を挙げるためには変数に取ったほうが読みやすくなることも多いので何が正解というのはないと思います。

個人的には変数に取らずにそのまま
(string)wSelectedRow.Cells["titleColumn"].Value
で渡してあげてもそこまで可読性は下がらないと思います。
ただもし、このシステムがもっと拡張されてwTitlwを利用したロジックが生まれる可能性が高いのであればこのままの方が正解の可能性もあります。

つまり何が言いたいかというと、「変数に取る」にも実は考えることが多いということ。

aizawa-being replied:
ありがとうございます！
今回、見やすければいいかと思い、変数に取る必要があるかは考えられていませんでした。
利用しない可能性がある変数を定義しても無駄なことをしているだけだと指摘されるまで考えられていませんでした。
少しのことに感じても大規模プロジェクトでは大きくなってしまうという観点も抜けていたため、変数一つだけと思わず、注意してなくてはならないと忘れないようにします。

k-kobayashi-being commented on Libra/Views/LibraForm.cs:
```diff
@@ -46,7 +46,49 @@ public partial class LibraForm : Form {
+            // 貸出した書籍にフォーカスします。
+            int wColumnIndex = this.booksDataGridView.Columns[0].Index;
+            foreach (DataGridViewRow wRow in this.booksDataGridView.Rows) {
+                if (wRow.Cells[wColumnIndex].Value != null && (int)wRow.Cells[wColumnIndex].Value == wBookId) {
```
ここは変数に取ってないもんねw

aizawa-being committed:
#438537 レンタル機能

余分な空白行の削除。取得する変数の見直し。著者名が存在しない書籍を取得した場合、（著者名なし）と登録するように修正。

aizawa-being committed:
#438537 レンタル機能

余分な空白行の削除。変数の宣言の見直し。

aizawa-being committed:
Merge branch 'feature/#438537_レンタル機能' of https://github.com/aizawa-being/Libra into feature/#438537_レンタル機能

aizawa-being committed:
#438537 レンタルシステム

コメントの修正。

aizawa-being committed:
Merge branch 'feature/#438537_レンタル機能' of https://github.com/aizawa-being/Libra into feature/#438537_レンタル機能

aizawa-being committed:
誤った行の修正。
